x-pack/plugins/task_manager/server/ephemeral_task_lifecycle.ts

```diff
@@ -0,0 +1,200 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Subject, Observable } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { Logger } from '../../../../src/core/server';
+
+import { Result, asErr, asOk } from './lib/result_type';
+import { TaskManagerConfig } from './config';
+
+import { asEphemeralTaskDelayedDueToCapacityEvent, TaskEventType } from './task_events';
+import { Middleware } from './lib/middleware';
+import { EphemeralTaskInstance } from './task';
+import { TaskTypeDictionary } from './task_type_dictionary';
+import { TaskLifecycleEvent } from './polling_lifecycle';
+import { EphemeralTaskManagerRunner } from './task_running/ephemeral_task_runner';
+import { TaskPool } from './task_pool';
+
+export interface EphemeralTaskLifecycleOpts {
+  logger: Logger;
+  definitions: TaskTypeDictionary;
+  config: TaskManagerConfig;
+  middleware: Middleware;
+  elasticsearchAndSOAvailability$: Observable<boolean>;
+  pool: TaskPool;
+  lifecycleEvent: Observable<TaskLifecycleEvent>;
+}
+
+export type EphemeralTaskInstanceRequest = Omit<EphemeralTaskInstance, 'startedAt'>;
+
+/**
+ * The public interface into the task manager system.
+ */
+export class EphemeralTaskLifecycle {
+  private definitions: TaskTypeDictionary;
+  private pool: TaskPool;
+  private lifecycleEvent: Observable<TaskLifecycleEvent>;
+  // all task related events (task claimed, task marked as running, etc.) are emitted through events$
+  private events$ = new Subject<TaskLifecycleEvent>();
+  private ephemeralTaskQueue = new Set<EphemeralTaskInstanceRequest>();
+  private logger: Logger;
+  private config: TaskManagerConfig;
+  private middleware: Middleware;
+
+  /**
+   * Initializes the task manager, preventing any further addition of middleware,
+   * enabling the task manipulation methods, and beginning the background polling
+   * mechanism.
+   */
+  constructor({
+    logger,
+    middleware,
+    definitions,
+    pool,
+    lifecycleEvent,
+    config,
+  }: EphemeralTaskLifecycleOpts) {
+    this.logger = logger;
+    this.middleware = middleware;
+    this.definitions = definitions;
+    this.pool = pool;
+    this.lifecycleEvent = lifecycleEvent;
+    this.config = config;
+
+    this.lifecycleEvent
+      .pipe(
+        filter(
+          // we want to know when the queue has ephemeral task run requests
+          (e) => {
+            return (
+              this.ephemeralTaskQueue.size > 0 &&
+              e.type === TaskEventType.TASK_POLLING_CYCLE &&
+              // and when a polling cycle has just completed,
+              // (e.type === TaskEventType.TASK_POLLING_CYCLE ||
+              //   // or the "load" in the TaskPool has changed (meaning a task has just completed)
+              //   (e.type === TaskEventType.TASK_MANAGER_STAT && e.id === 'load')) &&
+              this.getCapacity() === 0
```

Comment:
This will reject any ephemeral task enqueued when there's no capacity - why are we doing that? 🤨 
This is a very likely scenario... and is likely to happen a lot.
The idea would be to add those to the queue so they're executed once the running tasks are completed.

If the goal here was to make this a bounded queue, then that should be done synchronously in `attemptToRun` I think.

Reply:
Yes, this is legacy code that I'll remove. The idea behind this lifecycle hook is if capacity is at 0, send an event so the scheduler can reject the promise and the consumer can handle that rejection, which was previously causing the event to be queued in the current method (non ephemerally)